skunkworks_crow/src/main/java/org/odk/share/tasks/DownloadJob.java

```diff
@@ -94,26 +104,61 @@ protected Result onRunJob(@NonNull Params params) {
         ((Share) getContext().getApplicationContext()).getAppComponent().inject(this);
 
         initJob(params);
-        rxEventBus.post(receiveForms());
 
         return null;
     }
 
     private void initJob(Params params) {
-        ip = params.getExtras().getString(IP, "");
-        port = params.getExtras().getInt(PORT, -1);
         sbResult = new StringBuilder();
+        int method = params.getExtras().getInt("MODE_OF_TRANSFER", -1);
+        switch (method) {
+            case Share.TransferMethod.HOTSPOT:
+                ip = params.getExtras().getString(IP, "");
+                port = params.getExtras().getInt(PORT, -1);
+                break;
+            case Share.TransferMethod.BLUETOOTH:
+                targetMacAddress = params.getExtras().getString("mac", null);
+                break;
+        }
+
+        setupDataStreamsAndReceive(method, targetMacAddress);
```

Comment:
If you kept `targetMacAddress `as global then there is no need to pass.

Reply:
Fixed.